stl/inc/format

```diff
@@ -1205,69 +1216,181 @@ private:
+        _Handler::_On_sign(_Sgn);
+    }
+
     constexpr void _On_zero() {
-        // Note 0 is again not valid for CharT or bool unless a numeric
-        // presentation type is used.
-        _Numeric_checker._Require_numeric_argument();
+        _Need_arithmetic_presentation_type = true;
+        _Require_numeric_argument();
         _Handler::_On_zero();
     }
 
     constexpr void _On_precision(int _Precision) {
-        _Numeric_checker._Check_precision();
+        _Check_precision();
         _Handler::_On_precision(_Precision);
     }
+
+    template <class _Ty>
+    constexpr void _On_dynamic_precision(const _Ty _Val) {
+        _Check_precision();
+        _Handler::_On_dynamic_precision(_Val);
+    }
+
+    template <class _CharT>
+    constexpr void _On_type(_CharT _Type) {
+        if (_Type < 0 || _Type > (numeric_limits<signed char>::max) ()) {
+            _THROW(format_error("Invalid type specification."));
+        }
+        char _Narrow_type = static_cast<char>(_Type);
+        enum class _Presentation_type_category { _Default, _Integer, _Floating, _String, _Pointer, _Char_default };
+        auto _Cat = _Presentation_type_category::_Default;
+        switch (_Narrow_type) {
+        case '\0':
+            break;
+        case 's':
+            _Cat = _Presentation_type_category::_String;
+            break;
+        case 'c':
+        case 'd':
+        case 'B':
+        case 'b':
+        case 'X':
+        case 'x':
+        case 'o':
+            _Cat = _Presentation_type_category::_Integer;
+            break;
+        case 'A':
+        case 'a':
+        case 'E':
+        case 'e':
+        case 'F':
+        case 'f':
+        case 'G':
+        case 'g':
+            _Cat = _Presentation_type_category::_Floating;
+            break;
+        case 'p':
+            _Cat = _Presentation_type_category::_Pointer;
+            break;
+        default:
+            _THROW(format_error("Invalid presentation type specifier"));
+        }
+
+        switch (_Arg_type) {
+        case _Basic_format_arg_type::_None:
+            _STL_INTERNAL_CHECK(!"Invalid argument type.");
+            break;
+        case _Basic_format_arg_type::_Bool_type:
+            if (_Cat == _Presentation_type_category::_Default) {
+                _Cat = _Presentation_type_category::_String;
+            }
+            // note, we don't get a call if there isn't a type, but none is valid for everything.
+            if (_Cat != _Presentation_type_category::_String && _Cat != _Presentation_type_category::_Integer) {
+                _THROW(format_error("Invalid presentation type for bool"));
+            }
+            break;
+        case _Basic_format_arg_type::_Char_type:
+            if (_Cat == _Presentation_type_category::_Default) {
+                _Cat = _Presentation_type_category::_Char_default;
+            }
+            if (_Cat != _Presentation_type_category::_Char_default && _Cat != _Presentation_type_category::_Integer) {

```

Comment:
Convention: There should be a newline between non-chained `if` statements. Occurs below.